src/Store/VimStoreConnector.re

```diff
@@ -317,17 +319,51 @@ let start =
       open Vim.BufferUpdate;
       Log.info("Vim - Buffer update: " ++ string_of_int(update.id));
       open Core.Types;
+      open Model.State;
+
+      let isFull = update.endLine == (-1);
+
+      let bufferMaybe =
+        Some(getState())
+        |> Option.map(state => state.buffers)
+        |> Option.bind(Model.Buffers.getBuffer(update.id));
+
+      // If this is a 'full' update, check if there was a previous buffer.
+      // We need to keep track of the previous line count for some
+      // buffer synchronization strategies (ie, extension host)
+      let endLine =
+        if (isFull) {
+          bufferMaybe
+          |> Option.map(b => Model.Buffer.getNumberOfLines(b) + 1)
+          |> Option.value(~default=update.startLine);
+        } else {
+          update.endLine;
+        };
+
       let bu =
         Core.Types.BufferUpdate.create(
           ~id=update.id,
+          ~isFull,
           ~startLine=Index.OneBasedIndex(update.startLine),
-          ~endLine=Index.OneBasedIndex(update.endLine),
+          ~endLine=Index.OneBasedIndex(endLine),
           ~lines=update.lines,
           ~version=update.version,
           (),
         );
 
-      dispatch(Model.Actions.BufferUpdate(bu));
+      let shouldApply =
+        bufferMaybe
+        |> Option.map(Model.Buffer.shouldApplyUpdate(bu))
+        |> Option.value(~default=true);
+
+      if (shouldApply) {
```

Comment:
Add comment that we should be handling the buffer updates better from `reason-libvim` - in other words, we shouldn't _need_ this logic at this layer